[**&#10094; Previous**](a07_operacoes_matematicas.ipynb) | [**Home**](a00_aeds1.ipynb) | [**Next &#10095;**](a07_operadores_relacionais_e_logicos.ipynb)

# Operações matemáticas

Vimos [anteriormente](a07_operacoes_matematicas.ipynb) as seguintes operações matemáticas em C:

## Operações básicas:

In [1]:
#include <iostream>
#include <cmath>
using namespace std;

In [2]:
float f;

In [6]:
int i;

In [12]:
f = (10 + 5) / 20.0;
f++;
cout << f << endl;

f *= 2;
cout << f << endl;

i = 25 % 7;
cout << i << endl;

i = 25 / 7;
cout << i << endl;

1.75
3.5
4
3


* O uso dos parênteses força que a soma (10 + 5) aconteça antes da divisão;
* O uso do `float` 20.0 força o uso da divisão de números reais, e não a divisão de números inteiros;
* O operador `++` realiza o incremento de 1 na variável `f`;
* O operador `*=` multiplica `f` por 2 e salva o resultado na própria variável;
* O operador `%` retorna o resto da divisão entre 2 números inteiros;
* Na operação 25/7 é realizado a divisão de 2 números inteiros, e o resultado é um número inteiro.

## Outras operações matemáticas

Além das operações básicas, também temos disponível na biblioteca `cmath` diversas outras funções matemáticas, a saber:

## Constantes

Vamos definir 2 constantes para fazer a conversão de graus para radianos e vice-versa:

In [45]:
// definindo a constante DEG2RAD que converte um ângulo de graus para radianos (2*pi/360)
// definindo a constante RAD2DEG que converte um ângulo de radianos para graus (360/2*pi)
#define DEG2RAD 0.0174532925199433
#define RAD2DEG 57.2957795130823

## Declaração de variáveis

In [13]:
float theta;
float tangente;

## Corpo do código

In [74]:
theta = 30*DEG2RAD;

// funções de seno, cosseno e tangente:
cout << sin(theta) << endl;
cout << cos(theta) << endl;
cout << tan(theta) << endl;

// funções de seno hiperbólico, cosseno hiperbólico e tangente hiperbólica:
cout << sinh(theta) << endl;
cout << cosh(theta) << endl;
cout << tanh(theta) << endl;

// funções de arco seno, arco cosseno:
cout << asin(0.5)*RAD2DEG << endl;
cout << acos(0.5)*RAD2DEG << endl;

tangente = tan(theta);
// funções de arco tangente:
cout << atan(tangente)*RAD2DEG << endl;
cout << atan2(tangente,1)*RAD2DEG << endl;
cout << atan2(-tangente,-1)*RAD2DEG << endl;

0.5
0.866025388240814
0.577350318431854
0.547853469848633
1.1402382850647
0.480472832918167
30
60
30.000000834826
30.0000021160288
-149.999997883971


No código acima, vimos como chamar as funções trigonométricas em C, em especial, vale notar que:

* As funções trigonométricas trabalham sempre com ângulos fornecidos em radianos. É necessário realizar a conversão para graus sempre que necessário.
* A função arco tangente possui 2 implementações:
 * `atan(a)` que recebe apenas um argumento (a) e calcula o arco tangente de `a`.
 A função `atan` retorna um ângulo no intervalo de $-\pi/2$ à $\pi/2$ .
 * `atan2(y,x)` que recebe 2 argumentos, e calcula o arco tangente de y/x.
 A função `atan2` retorna um ângulo no intervalo de $-\pi$ à $\pi$ .


## Outras funções matemáticas

Veremos adiante outras funções matemáticas:

In [92]:
// e^2
f = exp(2);
cout << f << endl;

// 2^3
f = pow(2,3);
cout << f << endl;

// raiz quadrada
f = sqrt(36);
cout << f << endl;

// ln(e)
f = log(exp(1));
cout << f << endl;

// log_10(100)
f = log10(100);
cout << f << endl;

// arrendondamento, truncamento
f = ceil(12.5);
cout << f << endl;
f = ceil(-12.5);
cout << f << endl;

f = floor(14.5);
cout << f << endl;
f = floor(-14.5);
cout << f << endl;

// valor absoluto
f = fabs(-2.5);
cout << f << endl;

// resto usando ponto flutuante
f = fmod(9,2.5);
cout << f << endl;

7.38905620574951
8
6
1
2
13
-12
14
-15
2.5
1.5


## Exercício 1

Suponha que você queira fazer uma simulação referente a uma aplicação.
Você irá fazer um aporte inicial e o seu capital irá ser atualizado mensalmente, conforme a taxa fornecida pelo usuário.

A fórmula para calcular o montante final é [dado por](https://www.todamateria.com.br/juros-compostos/):

$M = C(1 + i)^t$

onde:

* M: montante final obtido ao final da aplicação;
* C: capital inicial investido;
* i: taxa de juros;
* t: período de tempo;

Crie um programa que leia os valores do capital inicial, taxa de juros e período de tempo e calcule o montante final obtido.

### Declaração de variáveis

### Corpo do algoritmo

## Exercício 2

Para este exercício considere os valores das massas atômicas:

$H = 1u$ e $O = 16u$

e considere $6.022 \times 10^{23}$ para a constante de Avogadro.

Escreva um programa que pergunte ao usuário a quantidade de água, dado em gramas, e determine a quantidade de móleculas de água presentes.

Consulte: https://mundoeducacao.bol.uol.com.br/quimica/mol.htm

### Declaração de variáveis

### Corpo do algoritmo

## Exercício 3

Sabendo que para queimar 1,0 litro de álcool são necessários aproximadamente 6.000 litros de ar e que um litro de ar pesa aproximadamente 1,29g, crie um programa que leia a quantidade de combustível abastecido e calcule a massa de ar consumida.

### Declaração de variáveis

### Corpo do algoritmo

## Exercício 4

Considere que a gasolina seja constituída apenas de 2,2,4-trimetil-pentano. 
Escreva um programa que leia a quantidade de gasolina abastecida por um veículo (em kg) e determine a quantidade de oxigênio consumido (em kg) e a quantidade de dióxido de carbono (em kg) que será lançada de volta na atmosfera.
Dados: 

C=12u; O=16u

$1 C_8H_{18} + \frac{25}{2} O_2 \rightarrow 8 CO_2 + 9 H_2O$

### Declaração de variáveis

### Corpo do algoritmo

## Exercício 5

Fonte: https://brasilescola.uol.com.br/matematica/aplicacoes-das-leis-trigonometricas-um-triangulo-seno-.htm

O construtor deseja calcular a distância do ponto A ao ponto C, pontos onde a ponte será construída, entretanto ele não possui nenhuma ferramenta que meça essa distância, mas ele conhece de matemática e teve a seguinte ideia. “Como eu possuo uma ferramenta que calcula ângulos, conseguirei determinar o comprimento desta ponte”. Com isso ele marcou um ponto B, calculou o ângulo BÂC que foi igual a 85°, caminhou até o ponto B, uma distância de 2km, e calculou o ângulo ABC obtendo um ângulo de 65°. O construtor acredita que com essas informações será possível calcular o comprimento da ponte.

![](figuras/triang1.jpg)

Sabendo que a lei dos senos nos diz que:

$
\frac{\overline{BC}}{\sin{\hat{A}}} = \frac{\overline{AB}}{\sin{\hat{C}}} = \frac{\overline{AC}}{\sin{\hat{B}}}
$

Escreva um programa que leia os ângulos ABC e BAC e a distância $\overline{AB}$ e determine o tamanho do segmento $\overline{AC}$.

### Declaração de variáveis

### Corpo do algoritmo

## Exercício 6

Suponha que desejemos medir a altura de um objeto (um prédio, uma montanha, etc).
A partir de um ponto C, utilizando um instrumento de medição de ângulo chamado "Estação Total", medimos o
ângulo $\alpha$. 
Afastando uma distância conhecida, por exemplo 50m, do ponto C marcamos o ponto "imaginário" D.
Do ponto D medimos o ângulo $\beta$, conforme figura abaixo:

![https://sca.profmat-sbm.org.br/sca_v2/get_tcc3.php?id=81959](figuras/oliveira2015.png)

A partir da geometria temos que:

$
\tan{\alpha} = \frac{h'}{x}
$ e 
$
\tan{\beta} = \frac{h'}{x + 50}
$

Isolando $x$ na primeira equação e manipulando a segunda equação obtemos a seguinte expressão:

$
h' = \frac{50\tan{\alpha} \tan{\beta} }{\tan{\alpha} - \tan{\beta} }
$

Sabendo que a estação total está a uma altura de 1,5m do chão, a altura final é dada por:

$
h = h' + 1.5
$

Fonte: [Oliveira, Joerk da Silva "Aplicações da trigonometria na matemática"](https://sca.profmat-sbm.org.br/sca_v2/get_tcc3.php?id=81959)

Escreva um programa que leia os dados fornecidos pela estação total e calcule a altura do objeto.


### Declaração de variáveis

### Corpo do algoritmo

## Exercício 7

Medir as distâncias de estrelas sempre foi um desafio, pois diferentemente das medições diretas feitas na Terra, não é possível levar uma regua até as estrelas.
Uma forma utilizada a muitos anos é conhecido como método da paralaxe.

Leia o poster [Teixeira, Helena "Aplicações da trigonometria na astronomia - Medição de distâncias pelo método da paralaxe", ](http://www.cienciaviva.pt/img/Poster_HTeixeira.pdf).

Em seguida, escreva um programa que calcule a distância de uma estrela.


### Declaração de variáveis

### Corpo do algoritmo

## Exercício - rebite

Leia o arquivo disponível em https://essel.com.br/cursos/material/01/EnsaioMateriais/ensa07.pdf

Crie um algoritmo que determine o número de rebites necessário para unir duas chapas de aço, sabendo que:

* a tensão de cisalhamento (TC) equivalente a 75\% da tensão de tração (T)
* a tensão de tração (T) suportada por cada rebite é 650 MPa
* o diâmetro (D) de cada rebite é de 4mm
* as chapas suportarão uma força cortante (F) fornecida pelo usuário

O número de rebites é dado por $n = \frac{F}{TC \cdot S}$, onde

$S = \frac{\pi D^2}{4}$ é a área da seção transversal do rebite.

### Declaração de constantes

### Declaração de variáveis

### Corpo do algoritmo

[**&#10094; Previous**](a07_operacoes_matematicas.ipynb) | [**Home**](a00_aeds1.ipynb) | [**Next &#10095;**](a07_operadores_relacionais_e_logicos.ipynb)